Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,252 @@
+    private var stream: UnsafeMutablePointer<Target>
+
+    public init(stream: UnsafeMutablePointer<Target>) {
+        if case let stdStream as OutputFileStreamProtocol = stream.pointee {
+            dumb = !(isatty(stdStream.fd) == 0)
+        } else {
+            dumb = true
+        }
+
+        self.stream = stream
+    }
+
+    /// Code to clear the line on a tty.
+    private let clearLineString = "\u{001B}[2K"
+
+    /// Code to end any currently active wrapping.
+    private let resetString = "\u{001B}[0m"
+
+    /// Clears the current line and moves the cursor to beginning of the line..
+    public func clearline() {
+        guard !dumb else { return }
+        print(clearLineString + "\r", terminator: "", to: &stream.pointee)
+    }
+
+    /// Moves the cursor y columns up.
+    public func moveCursor(y: Int) {
+        guard !dumb else { return }
+        print("\u{001B}[\(y)A", terminator: "", to: &stream.pointee)
+    }
+
+    /// Writes a line to the output stream. Truncates the line in middle if line is longer than the terminal width.
+    /// If not writing to a terminal entire line will be written without truncation.
+    public func writeLine(line: String, color: TermColor = .noColor) {
+        let wrappedLine = wrap(line, inColor: color)
+        print(truncate(wrappedLine), terminator: "", to: &stream.pointee)
+        flush()
+    }
+
+    /// Writes a string to the stream.
+    public func write(string: String, color: TermColor = .noColor) {
+        print(wrap(string, inColor: color), terminator: "", to: &stream.pointee)
+        flush()
+    }
+
+    /// Inserts a new line character into the stream.
+    public func endLine() {
+        print("\n", terminator: "", to: &stream.pointee)
+        flush()
+    }
+
+    /// Flushes the stream if writing to a filestream.
+    public func flush() {
+        if case let stdStream as OutputFileStreamProtocol = stream.pointee {
+            fflush(stdStream.file)
+        }
+    }
+
+    /// Truncates the given line in middle to fit the terminal width if writing to tty
+    /// otherwise returns the same string.
+    private func truncate(_ line: String) -> String {
+        // Return the original line if we don't need to truncate.
+        guard !dumb && line.utf8.count > terminalWidth else {
+            return line
+        }
+        let mid = terminalWidth/2
+        let frontIndex = line.index(line.startIndex, offsetBy: max(0, mid - 2))
+        let backIndex = line.index(line.endIndex, offsetBy: -(mid + 1))
+        return line[line.startIndex..<frontIndex] + "..." + line[backIndex..<line.endIndex]
+    }
+
+    /// Wraps the string into the color mentioned. If terminal is dumb no color code will be inserted.
+    private func wrap(_ line: String, inColor color: TermColor) -> String {
+        guard !dumb || !line.isEmpty || color == .noColor else {
+            return line
+        }
+        return color.string + line + resetString
```

Comment:
Actually... I might take this back.

I wonder if clients need this function to be public? The reason why is that I suspect it might make sense in places to form a complete line before writing it (this is definitely true if the API embeds the `flush`) so that it draws as nicely as possible on a terminal. I remember tuning how the writes were done in `llbuild` to get cleaner output (less flashing) and clients should have control over that.